source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -316,6 +304,111 @@ void RedisCluster::RedisDiscoverySession::startResolveRedis() {
+            if (status != Network::DnsResolver::ResolutionStatus::Success) {
+              ENVOY_LOG(error, "Unable to resolve cluster slot primary hostname {}",
+                        slot.primary_hostname_);
+              resolve_timer_->enableTimer(parent_.cluster_refresh_rate_);
+              return;
+            }
+            // Primary slot address resolved
+            slot.setPrimary(Network::Utility::getAddressWithPort(
+                *response.front().addrInfo().address_, slot.primary_port_));
+            // Continue on to resolve replicas
+            resolveReplicas(slots, slot_idx);
+          });
+    } else {
+      resolveReplicas(slots, slot_idx);
+    }
+    slot_idx++;
+  }
+}
+
+/**
+ * Resolve the replicas in a cluster entry. If there are no replicas, simply move to the next
+ * cluster entry. Once the last replica is resolved, we move to resolving the next cluster entry.
+ *
+ * @param slots the list of slots which may need DNS resolution
+ * @param index the specific index into `slots` whose replicas need to be resolved
+ */
+void RedisCluster::RedisDiscoverySession::resolveReplicas(ClusterSlotsSharedPtr slots,
+                                                          std::size_t index) {
+  auto& slot = (*slots)[index];
+  if (slot.replicas_to_resolve_.empty()) {
+    if (index == (*slots).size() - 1) {
+      finishClusterHostnameResolution(slots);
+    }
+    return;
+  }
+
+  std::size_t replica_idx = 0;
+  while (replica_idx < slot.replicas_to_resolve_.size()) {
+    auto replica = slot.replicas_to_resolve_[replica_idx];
+    ENVOY_LOG(trace, "starting async DNS resolution for replica address {}", replica.first);
+    parent_.dns_resolver_->resolve(
+        replica.first, parent_.dns_lookup_family_,
+        [this, index, slots, replica_idx](Network::DnsResolver::ResolutionStatus status,
+                                          std::list<Network::DnsResponse>&& response) -> void {
+          auto& slot = (*slots)[index];
+          auto& replica = slot.replicas_to_resolve_[replica_idx];
+          ENVOY_LOG(trace, "async DNS resolution complete for {}", replica.first);
+          updateDnsStats(status, response.empty());
+          // If DNS resolution fails here, we move on to resolve other replicas in the list.
+          // We log a warn message.
+          if (status != Network::DnsResolver::ResolutionStatus::Success) {
+            ENVOY_LOG(warn, "Unable to resolve cluster replica address {}", replica.first);
+          } else {
+            // Replica resolved
+            slot.addReplica(Network::Utility::getAddressWithPort(
+                *response.front().addrInfo().address_, replica.second));
+          }
+          // finish resolution if it's the last primary and last replica of that primary being
+          // resolved
+          if (index == (*slots).size() - 1 && replica_idx == slot.replicas_to_resolve_.size() - 1) {
+            finishClusterHostnameResolution(slots);
```

Comment:
Since we're doing these in parallel now don't we need to wait for all of them to finish? Not just the last one?